middle_end/flambda/simplify/simplify_extcall.ml

```diff
@@ -0,0 +1,159 @@
+  | Poly_compare_specialized of DA.t * Expr.t
+
+(* Helpers *)
+(* ******* *)
+
+let fun_symbol simple =
+  let fail simple =
+    Misc.fatal_errorf
+      "Expected a function symbol, instead of@ %a" Simple.print simple
+  in
+  Simple.pattern_match simple
+    ~name:(fun name ->
+      Name.pattern_match name
+        ~var:(fun _ -> fail simple)
+        ~symbol:(fun sym -> sym))
+    ~const:(fun _ -> fail simple)
+
+let simple_of_var v =
+  Simple.name (Name.var v)
+
+let apply_cont cont v ~dbg =
+  let args = [ Simple.name (Name.var v) ] in
+  let apply_cont = Apply_cont.create cont ~args ~dbg in
+  let free_names = Apply_cont.free_names apply_cont in
+  let expr = Expr.create_apply_cont apply_cont in
+  free_names, expr
+
+let let_prim ~dbg v prim (free_names, body) =
+  let v' = Var_in_binding_pos.create v Name_mode.normal in
+  let bindable = Bindable_let_bound.singleton v' in
+  let named = Named.create_prim prim dbg in
+  let free_names_of_body = Or_unknown.Known free_names in
+  let let_expr = Let.create bindable named ~body ~free_names_of_body in
+  let free_names = Name_occurrences.remove_var free_names v in
+  let expr = Expr.create_let let_expr in
+  free_names, expr
+
+
+(* Exported simplification function *)
+(* ******************************** *)
+
+let simplify_comparison ~dbg ~dacc ~cont
+      ~int_prim ~float_prim a b a_ty b_ty =
+  let tenv = DA.typing_env dacc in
+  match T.prove_is_a_tagged_immediate tenv a_ty,
+        T.prove_is_a_tagged_immediate tenv b_ty with
+  | Proved (), Proved () ->
+    let v_comp = Variable.create "comp" in
+    let tagged = Variable.create "tagged" in
+    let _, res =
+      let_prim ~dbg v_comp (P.Binary (int_prim, a, b)) @@
+      let_prim ~dbg tagged
+        (P.Unary (Box_number Untagged_immediate, simple_of_var v_comp)) @@
+      apply_cont ~dbg cont tagged
+    in
+    Poly_compare_specialized (dacc, res)
+  | _ ->
+    begin match T.prove_is_a_boxed_float tenv a_ty,
+                T.prove_is_a_boxed_float tenv b_ty with
+    | Proved (), Proved () ->
+      let a_naked = Variable.create "naked_float" in
+      let b_naked = Variable.create "naked_float" in
+      let v_comp = Variable.create "comp" in
+      let tagged = Variable.create "tagged" in
+      let _, res =
+        let_prim ~dbg a_naked (P.Unary (Unbox_number Naked_float, a)) @@
+        let_prim ~dbg b_naked (P.Unary (Unbox_number Naked_float, b)) @@
+        let_prim ~dbg v_comp
+          (P.Binary (float_prim, simple_of_var a_naked, simple_of_var b_naked)) @@
+        let_prim ~dbg tagged
+             (P.Unary (Box_number Untagged_immediate, simple_of_var v_comp)) @@
+        apply_cont ~dbg cont tagged
+      in
+      Poly_compare_specialized (dacc, res)
+    | _ -> Unchanged
+    end
+
+
+let simplify_returning_extcall
+      ~dbg ~cont ~exn_cont:_ dacc fun_name args ~arg_types =
+  match fun_name, args, arg_types with
+  (* Polymorphic comparisons *)
+  | ".extern__caml_compare", [a; b], [a_ty; b_ty] ->
```

Comment:
This string matching is fragile. We'd like to be able to catch a name change. (and the removal of `.extern__`)
This can be tested by making a test of the form: (for each kind of comparison)
```
let[@inline] test x y =
  let res = if Obj.magic (x < y) < 2 then 1 else Sys.opaque_identity 42 in
  Some res

let[@inline never] test2 (x:float) y = test x y
```
If the transformation is done, the primitive is known and the type can tell that it's below 2 hence `Some 42` is a constant. So no allocations

Reply:
Test added (to cover all comparison functions) in the testsuite